File: Python/nlp_to_signal.py
# -*- coding: utf-8 -*-
"""
NLP to Signal Processing Lab (Python)
-------------------------------------
Goal: Use NLP to generate and analyze signals.
"""

# %% [markdown]
# # Part 1: Signal Generation from Text
# In this section, we define a simple function that parses a natural language 
# description to generate a signal.

import numpy as np
import matplotlib.pyplot as plt
from scipy import signal
import re

def text_to_signal(text, fs=1000, duration=1.0):
    """
    Simple rule-based 'NLP' to extract frequency and type.
    In a real lab, you'd use a transformer model here.
    """
    # Extract frequency using regex
    freq_match = re.search(r'(\d+)\s*Hz', text, re.IGNORECASE)
    freq = float(freq_match.group(1)) if freq_match else 10.0
    
    t = np.linspace(0, duration, int(fs * duration), endpoint=False)
    
    if 'sine' in text.lower():
        sig = np.sin(2 * np.pi * freq * t)
    elif 'square' in text.lower():
        sig = signal.square(2 * np.pi * freq * t)
    else:
        sig = np.sin(2 * np.pi * freq * t) # Default
        
    if 'noise' in text.lower():
        sig += 0.5 * np.random.randn(len(t))
        
    return t, sig

# Example usage
description = "A 50Hz square wave with noise"
t, sig = text_to_signal(description)

plt.figure(figsize=(10, 4))
plt.plot(t[:200], sig[:200])
plt.title(f"Generated Signal: {description}")
plt.grid(True)
plt.show()

# %% [markdown]
# # Part 2: Automated Filtering
# Let's say we want to "Remove high frequency noise".

def apply_nlp_filter(sig, fs, instruction):
    if "remove high frequency" in instruction.lower() or "low pass" in instruction.lower():
        # Design a 5th order Butterworth filter
        sos = signal.butter(10, 100, 'lp', fs=fs, output='sos')
        filtered = signal.sosfilt(sos, sig)
        return filtered
    return sig

instruction = "Apply a low pass filter to remove noise"
filtered_sig = apply_nlp_filter(sig, 1000, instruction)

plt.figure(figsize=(10, 4))
plt.plot(t[:200], sig[:200], alpha=0.5, label='Original')
plt.plot(t[:200], filtered_sig[:200], label='Filtered')
plt.legend()
plt.title("NLP Controlled Filtering")
plt.show()

# %% [markdown]
# # Part 3: Using a Real LLM (Instructional)
# To use a real LLM (like Gemini or GPT-4), you would send a prompt like:
# "Generate Python code using scipy to design a bandpass filter from 20Hz to 50Hz 
# for a signal sampled at 1kHz."

File: MATLAB/nlp_to_signal.m
%% NLP to Signal Processing Lab (MATLAB)
% Goal: Explore MATLAB's capabilities in NLP-driven Signal Processing.

%% Part 1: Basic Signal Synthesis from Description
% MATLAB's high-level functions make it easy to map text parameters to signals.

description = "50Hz sine wave with 0.2 amplitude noise";

% Simple parsing (Equivalent to the Python lab)
freq = 50; 
fs = 1000;
t = 0:1/fs:1-1/fs;

if contains(description, 'sine', 'IgnoreCase', true)
    sig = sin(2*pi*freq*t);
end

if contains(description, 'noise', 'IgnoreCase', true)
    sig = sig + 0.2*randn(size(t));
end

figure;
plot(t(1:200), sig(1:200));
title(['Generated: ', description]);
grid on;

%% Part 2: Interactive Filtering with Signal Analyzer
% MATLAB provides the "Signal Analyzer" app which can be controlled via 
% scripts. 

% Task: Use the following code to design a filter based on 'instructions'
instruction = "low pass filter at 100Hz";

if contains(instruction, 'low pass')
    [b, a] = butter(6, 100/(fs/2));
    filteredSig = filter(b, a, sig);
end

hold on;
plot(t(1:200), filteredSig(1:200), 'LineWidth', 1.5);
legend('Original', 'Filtered');
title('NLP-Based Filtering in MATLAB');

%% Part 3: Bridge to Python NLP
% MATLAB can call Python functions directly. If you have the Python 
% 'transformers' library installed, you can use it here.

try
    % Example of calling Python's regex from MATLAB as a proxy for NLP
    pyRegex = py.importlib.import_module('re');
    match = pyRegex.search('(\d+)Hz', description);
    extractedFreq = char(match.group(py.int(1)));
    fprintf('Extracted Frequency via Python NLP: %s Hz\n', extractedFreq);
catch
    disp('Python not configured or library missing. Skipping bridge example.');
end

%% Part 4: Using MATLAB's AI Chat Playground (Manual Step)
% 1. Open the "AI Chat Playground" in MATLAB (R2024a+).
% 2. Type: "Generate a spectrogram for a chirp signal with noise."
% 3. Observe the code generation and execution.
